bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/ThingDiscoveryService.java

```diff
@@ -0,0 +1,244 @@
+        logger.debug("SHC has {} devices", devices.size());
+
+        // Write found devices into openhab.log to support manual configuration
+        for (Device d : devices) {
+            logger.info("Found device: name={} id={}", d.name, d.id);
+            if (d.deviceServiceIds != null) {
+                for (String s : d.deviceServiceIds) {
+                    logger.info(".... service: {}", s);
+                }
+            }
+        }
+
+        addDevices(devices, rooms);
+    }
+
+    protected void addDevices(List<Device> devices, List<Room> rooms) {
+        for (Device device : devices) {
+            addDevice(device, getRoomNameForDevice(device, rooms));
+        }
+    }
+
+    protected String getRoomNameForDevice(Device device, List<Room> rooms) {
+        return rooms.stream().filter(room -> room.id.equals(device.roomId)).findAny().map(r -> r.name).orElse("");
+    }
+
+    protected void addDevice(Device device, String roomName) {
+        // see startScan for the runtime null check of shcBridgeHandler
+        assert shcBridgeHandler != null;
+
+        logger.trace("Discovering device {}", device.name);
+        logger.trace("- details: id {}, roomId {}, deviceModel {}", device.id, device.roomId, device.deviceModel);
+
+        ThingTypeUID thingTypeUID = getThingTypeUID(device);
+        if (thingTypeUID == null) {
+            return;
+        }
+
+        logger.trace("- got thingTypeID '{}' for deviceModel '{}'", thingTypeUID.getId(), device.deviceModel);
+
+        ThingUID thingUID = new ThingUID(thingTypeUID, shcBridgeHandler.getThing().getUID(),
+                device.id.replace(':', '_'));
+
+        logger.trace("- got thingUID '{}' for device: '{}'", thingUID, device);
+
+        DiscoveryResultBuilder discoveryResult = DiscoveryResultBuilder.create(thingUID).withThingType(thingTypeUID)
+                .withProperty("id", device.id).withLabel(getNiceName(device.name, roomName));
+        if (null != shcBridgeHandler) {
+            discoveryResult.withBridge(shcBridgeHandler.getThing().getUID());
+        }
+        if (!roomName.isEmpty()) {
+            discoveryResult.withProperty("Location", roomName);
+        }
+        thingDiscovered(discoveryResult.build());
+
+        logger.debug("Discovered device '{}' with thingTypeUID={}, thingUID={}, id={}, deviceModel={}", device.name,
+                thingUID, thingTypeUID, device.id, device.deviceModel);
+    }
+
+    private String getNiceName(String name, String roomName) {
+        // remove "-" from "-IntrusionDetectionSystem-"
+        String niceName = name.replace("-", "");
+        // convert "IntrusionDetectionSystem" into "Intrusion Detection System"
+        String[] allParts = StringUtils.splitByCharacterTypeCamelCase(niceName);
+        String[] parts = Arrays.stream(allParts).filter(p -> !p.trim().isEmpty()).collect(Collectors.toList())
+                .toArray(new String[0]);
```

Comment:
Array conversion is not needed. You could use `String.join()` providing a list of strings. But there is an even better way: use `Collectors.joining(" ")` instead of `Collectors.toList()` and you're done directly 👍